Move all flow control items into `flow control` group. Add `if let` explanations.

mdinger opened the pull request:
This isn't done but I'm working on moving all flow control structures together into the same folder and adding `if let` and `while let` which I don't think are explained anywhere except for their RFC.

Previously the flow controllers have been together mostly except for `match` which seems to be alone so some data structures can be introduced. Currently, I think it would be best to move the flow control together, explain them, and link to necessary external sources as necessary. This is not ideal but not bad either. I don't think theres a good reason the controllers should be separated.

In the future, it may be a good idea for the datatypes to all be moved to the beginning before the flow control. They shouldn't be as problematic to explain (except for `String`/`&str`).

I'm not sure moving the controllers around will be well received so I'm trying to explain a little beforehand. I'll try to get `while let` explanations up maybe tomorrow...we'll see.

I didn't spend a ton of time on the `if let` example. Mainly tried to get it to work. It may need tweaking.

cc @GBGamer You may be interested in this. Comments welcome, particularly with `if let`.

rust-highfive commented:
r? @steveklabnik

(rust_highfive has picked a reviewer for you, use r? to override)

steveklabnik commented on examples/flow_control/if-else/input.md:
```diff
@@ -4,3 +4,6 @@ block. If-else conditionals are expressions too; and, because of Rust type
 
 {if-else.play}
+
+###See also:
```
I think there should be a space between the `#` and the `S`

mdinger replied:
Fixed all these.

steveklabnik commented on examples/flow_control/match/guard/guard.rs:
```diff
@@ -3,13 +3,10 @@ fn main() {
         (x, y) if x == y => println!("These are twins"),
         // The ^ `if condition` part is a guard
         (x, y) if x + y == 0 => println!("Antimatter, kaboom!"),
-        // `_` means don't bind the value to a variable
```
Why remove this stuff?

mdinger replied:
I split this into 2 sections because the way it was previously named "Destructuring and guards". The problem was when in a deeply nested folder (`Flow_control -> Pattern matching -> Destructuring and guards`), `Destructuring and guards` didn't fit in the sidebar. So I split it into 2 sections: `Destructuring` and `Guards`.

`Destructuring` (directly before `Guards`) covers those items so I dropped them. I'm not sure I split it in the most ideal way but I'll look at it again later. Opinions welcome.

mdinger commented on examples/flow_control/match/destructure/input.md:
```diff
@@ -0,0 +1,6 @@
+{destructure.play}
+
+###See also:
+[Tuples](/tuples.html)
```
I didn't know links could use the local (`/tuples.html`) syntax until recently. I thought they needed to be absolute. This is nice.

mdinger commented:
Added `while let` and reworked the `if let` example a bit. I think it's better. I've gotta let it sit for a bit so I can try and come look the whole thing over fresh later.

mdinger commented on examples/flow_control/if_let/input.md:
```diff
@@ -0,0 +1,30 @@
+[`enum`][enum], [`Option`][option], and the [RFC][if_let_rfc]
+
+[enum]: /enum.html
+[if_let_rfc]: https://github.com/rust-lang/rfcs/pull/160
```
Added RFC links for  `if let` and `while let`. May drop it. Kinda useful for more info if someones interested. Uncertain.

mdinger commented on examples/flow_control/while_let/input.md:
````diff
@@ -0,0 +1,39 @@
+}
+```
+
+`while let` makes this sequence much nicer:
````
Note for me to remember: `while let` does *not* have the corresponding `else`/`else if` extensions that `if let` has and I probably should comment about it.

mdinger replied:
Fixed

mdinger commented:
Added more examples covering `ref`, pointer, enum destructuring, and `match` binding. Gotta look over it again but it seems much more comprehensive than it was before.

@steveklabnik I don't know if you're gonna want to review this in one patch...or what. Definitely easier to peruse if you make the examples locally though.

mdinger commented on examples/flow_control/match/destructuring/destructure_enum/input.md:
```diff
@@ -0,0 +1,10 @@
+An `enum` is destructured similarly:
```
Filenames are really long because some of the `id`s (like `struct`) that were being used in JSON were clashing with other locations (like `struct`). This fixed it and will help with future clashing but it's not really pretty.

mdinger commented:
r? @steveklabnik Made a few more minor changes. I definitely think this is an improvement over before (that is before the PR).

Trying to add cross referencing to important concepts as I go. Avoided linking to `println!` because that would add links to every page. Though, I don't mind doing so (later PR though if desired).

mdinger commented:
Made changes to elaborate on destructuring for #390

steveklabnik commented:
I've looked at this patch once or twice already and am just not 100% sure. I'm trying to get to rust by example later this wekk.

mdinger commented:
Okay. Just let me know what some of the reservations are when you get to it. Theoretically, it's possible they could be addressed but I'm also hesitant to make the patch bigger because review time gets bigger. Easier to fix later.

GBGamer commented:
@mdinger If you need any help with the patch still, tell me. Sorry I've been out of contact, I just got a new guitar and I've been playing it... a lot.

mdinger commented:
@GBGamer Thanks. I'm not anticipating many issues here except that in moving all the flow controls together, some things are now introduced too early. Pointers, structs, enums, and options are now used before they are introduced. So I tried to be liberal with cross referencing.

I don't think that something that can be avoided long term though. When looking at a language for the first time, people want to know what the structures, types, flow controls, and operators look like and don't want to have to hunt around for them. Not being explained (`option` for example) at first sight isn't a huge problem as long as you point to where they need to look though.

This depends on the review though. @steveklabnik may reject it without shuffling of structures also. Then I'd need an additional patch for that. You could help with that. But it entirely depends on what he says. He may also have additional concerns I haven't thought of.

My preference would be to merge and have the rest (possibly) restructured later. But I gotta wait I guess.

mdinger commented:
@steveklabnik This needs a rebase and updating but I'm hoping you'll give comments for and against this before I do that. I doubt I can update this until late next week.

steveklabnik commented:
@mdinger well Monday -> Weds I am flying from Europe to Australia, soooooo

mdinger commented:
Okay. I guess it can't be helped.

steveklabnik commented:
Okay! So. I think that this is good to merge, other than needing a rebase. @mdinger , please rebase it, i'll merge, and we can improve from there. Sorry about the continued delays :/

mdinger commented:
Okay cool!!

mdinger commented:
This threw errors with the new link checker but the links do work. I didn't track it down. It might be that the new folders on rustbyexample.com don't exist yet or that the link checker doesn't deal with a root directory properly (`/new_example`). I'm not sure.

mdinger commented:
@dbbrian See previous comment.

mdinger commented on examples/flow_control/match/destructuring/destructure_structures/input.md:
```diff
@@ -0,0 +1,6 @@
+{struct.play}
+
+### See also:
+[Structs](/structs.html)
```
The link checker is failing on links like this though they do work. They will only work when it's on rustbyexample.com or `make serve` is currently running though.

dbbrian commented:
@mdinger Will fix ASAP.

steveklabnik commented:
Awesome, glad to finally land this. thanks for your patience, @mdinger , and thanks for the quick link checker fix @dbbrian !

mdinger commented:
I'm soooooooo glad it finally landed. I was beginning to think the only non-trivial things that will ever land are the ones you write. Makes it hard to want to contribute ...

I really tried not to nag but seems like I had too...don't like doing it. I don't know. The docs still need so much improvement. Best to keep working to make them better.

Really glad you finally merged!!!

steveklabnik commented:
Such is open source. Sometimes you have to nag. It happens.
